File: weave_introduction_handson/en/1_3_multimodal_openai.py
"""
1_3: Multimodal Data Tracing - Tracing multimodal data

What you'll learn in this script:
================================
1. Adding media types with Annotated[bytes, Content[Literal[...]]]
2. Tracing PNG/JPG images
3. Tracing MP3 audio
4. Tracing MP4 video
5. Tracing PDFs
6. Tracing HTML

Where to look after running:
================================
- Traces tab: Preview each media payload
- Inputs/Outputs: Inspect byte payloads stored with Content types
"""

import io
import os
import base64
import tempfile
import urllib.request
from typing import Annotated, Literal

from dotenv import load_dotenv
from openai import OpenAI
from PIL import Image, ImageDraw
import weave
from weave import Content

from config_loader import init_weave

# Load environment variables
load_dotenv()

# Initialize Weave
init_weave()

# OpenAI client
client = OpenAI()


def download_bytes(url: str, timeout: int = 60) -> bytes:
    """Small helper that downloads bytes from a URL."""
    with urllib.request.urlopen(url, timeout=timeout) as response:
        return response.read()


# =============================================================================
# 1. Image (PNG) - OpenAI gpt-image-1
# =============================================================================
print("\n" + "=" * 60)
print("1. Image (PNG) - OpenAI gpt-image-1")
print("=" * 60)


@weave.op()
def generate_image(prompt: str) -> Annotated[bytes, Content[Literal["png"]]]:
    """Generate an image with OpenAI and log it as PNG bytes in Weave."""
    response = client.images.generate(
        model="gpt-image-1",
        prompt=prompt,
        size="1024x1024",
    )
    return base64.b64decode(response.data[0].b64_json)


image_bytes = generate_image("a small robot watering a flower in a lab, pixel art style")
print(f"PNG image payload size: {len(image_bytes):,} bytes")


# =============================================================================
# 2. Image (JPG) - Trace an existing image
# =============================================================================
print("\n" + "=" * 60)
print("2. Image (JPG) - Trace an existing image")
print("=" * 60)


@weave.op()
def create_sample_jpg_image() -> Annotated[bytes, Content[Literal["jpg"]]]:
    """Generate a sample JPG image so it can be previewed in Weave UI."""
    image = Image.new("RGB", (640, 360), color=(244, 247, 251))
    draw = ImageDraw.Draw(image)
    draw.rectangle((0, 0, 640, 90), fill=(31, 64, 104))
    draw.ellipse((72, 126, 252, 306), fill=(255, 195, 0), outline=(31, 64, 104), width=6)
    draw.rounded_rectangle((330, 130, 568, 300), radius=24, fill=(82, 167, 255), outline=(31, 64, 104), width=6)
    draw.line((92, 238, 232, 166), fill=(31, 64, 104), width=5)
    draw.line((366, 166, 530, 264), fill=(255, 255, 255), width=8)

    buffer = io.BytesIO()
    image.save(buffer, format="JPEG", quality=90)
    return buffer.getvalue()


jpg_bytes = create_sample_jpg_image()
print(f"JPG image size: {len(jpg_bytes):,} bytes")


# =============================================================================
# 3. Audio (MP3) - OpenAI TTS
# =============================================================================
print("\n" + "=" * 60)
print("3. Audio (MP3) - OpenAI TTS")
print("=" * 60)


@weave.op()
def text_to_speech(text: str, voice: str = "alloy") -> Annotated[bytes, Content[Literal["mp3"]]]:
    """Generate audio with OpenAI TTS and log it as MP3 bytes in Weave."""
    response = client.audio.speech.create(
        model="gpt-4o-mini-tts",
        voice=voice,
        input=text,
        response_format="mp3",
    )
    if hasattr(response, "read"):
        return response.read()
    if hasattr(response, "content"):
        return response.content
    return b"".join(response.iter_bytes())


audio_bytes = text_to_speech(
    "Weights and Biases Weave helps you trace, evaluate, and monitor LLM applications."
)
print(f"MP3 audio size: {len(audio_bytes):,} bytes")


# =============================================================================
# 4. Video (MP4) - OpenAI Sora 2
# =============================================================================
print("\n" + "=" * 60)
print("4. Video (MP4) - OpenAI Sora 2")
print("=" * 60)


@weave.op()
def generate_video(
    prompt: str = "A small robot waters a flower in a bright lab, cinematic lighting",
) -> Annotated[bytes, Content[Literal["mp4"]]]:
    """Generate a video with OpenAI Sora 2 and log it as MP4 bytes in Weave."""
    reference_image_url = "https://assets.st-note.com/img/1762403176-PimhEZu3voSeGp5C0l9t4z2N.png"
    png = download_bytes(reference_image_url, timeout=30)

    video = client.videos.create(
        model="sora-2-pro",
        prompt=prompt,
        input_reference=("ref.png", io.BytesIO(png), "image/png"),
        size="1280x720",
        seconds="12",
    )
    result = client.videos.poll(video.id)
    if result.status != "completed":
        raise RuntimeError(f"Video generation failed: {result.error}")

    content = client.videos.download_content(video.id)
    with tempfile.NamedTemporaryFile(suffix=".mp4", delete=False) as tmp:
        content.write_to_file(tmp.name)
        with open(tmp.name, "rb") as f:
            return f.read()


if os.getenv("RUN_EXPENSIVE_MULTIMODAL") == "1":
    video_bytes = generate_video()
    print(f"MP4 video size: {len(video_bytes):,} bytes")
else:
    print("Set RUN_EXPENSIVE_MULTIMODAL=1 to run the Sora 2 video generation example.")


# =============================================================================
# 5. PDF - arXiv paper
# =============================================================================
print("\n" + "=" * 60)
print("5. PDF - arXiv paper")
print("=" * 60)


@weave.op()
def fetch_arxiv_pdf(paper_id: str = "1706.03762") -> Annotated[bytes, Content[Literal["pdf"]]]:
    """Download an arXiv PDF and log it as PDF bytes in Weave."""
    return download_bytes(f"https://arxiv.org/pdf/{paper_id}.pdf", timeout=60)


pdf_bytes = fetch_arxiv_pdf()
print(f"PDF size: {len(pdf_bytes):,} bytes")


# =============================================================================
# 6. HTML - Web page
# =============================================================================
print("\n" + "=" * 60)
print("6. HTML - Web page")
print("=" * 60)


@weave.op()
def fetch_html(url: str = "https://example.com") -> Annotated[bytes, Content[Literal["html"]]]:
    """Return HTML as bytes so it can be previewed in Weave UI."""
    return download_bytes(url, timeout=30)


html_bytes = fetch_html()
print(f"HTML size: {len(html_bytes):,} bytes")


print("\n" + "=" * 60)
print("Multimodal Data Tracing Demo Complete!")
print("=" * 60)
print("""
Summary:
- Annotated[bytes, Content[Literal["png"]]]: PNG image
- Annotated[bytes, Content[Literal["jpg"]]]: JPG image
- Annotated[bytes, Content[Literal["mp3"]]]: MP3 audio
- Annotated[bytes, Content[Literal["mp4"]]]: MP4 video
- Annotated[bytes, Content[Literal["pdf"]]]: PDF
- Annotated[bytes, Content[Literal["html"]]]: HTML

Check in Weave UI:
- Use the Traces tab to preview images, audio, video, PDFs, and HTML
- Use Inputs/Outputs to inspect payloads stored with Content types
""")

File: weave_introduction_handson_typescript/jp/1_3_multimodal_openai.ts
/**
 * 1_3: Multimodal Data Tracing - マルチモーダルデータのトレース
 *
 * TypeScript 対応: 一部対応
 *
 * Python 版では `Annotated[bytes, Content[Literal["png"]]]` のように、
 * bytes に media type を付けて Weave UI で preview できます。
 *
 * TypeScript SDK では現時点で公式 helper が次に限られます。
 * - weave.weaveImage({ data }) -> PNG
 * - weave.weaveAudio({ data, audioType: "wav" }) -> WAV
 *
 * JPG / MP3 / MP4 / PDF / HTML は通常 object として metadata trace はできますが、
 * Python の Content annotation と同じ media preview としては扱っていません。
 *
 * 対応状況:
 * - PNG image: Python 対応 / TypeScript 対応 (weaveImage)
 * - JPG image: Python 対応 / TypeScript 未対応 (preview helper なし)
 * - MP3 audio: Python 対応 / TypeScript 未対応 (preview helper なし)
 * - WAV audio: Python 版 1_3 では未使用 / TypeScript 対応 (weaveAudio)
 * - MP4 video: Python 対応 / TypeScript 未対応 (preview helper なし)
 * - PDF: Python 対応 / TypeScript 未対応 (preview helper なし)
 * - HTML: Python 対応 / TypeScript 未対応 (preview helper なし)
 */

import * as weave from "weave";

import {
  getImageModelName,
  getOpenAIClient,
  getTtsModelName,
  initWeave,
  printSection,
} from "../src/config.js";

await initWeave();

const SAMPLE_VISIBLE_PNG = Buffer.from(
  "iVBORw0KGgoAAAANSUhEUgAAAUAAAAC0CAYAAADl5PURAAADbUlEQVR42u3UMQ3DQBBFwWMSKsYQFGEWIKZkyUWKCwCXdixFb4oh8Hf1xmN5TYCiYQRAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQEEBDAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCPDTAG77ZwIUCSAggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAAC+OfmOi5jTxDAVPTEEARQ9MQQBFD4hBAEUPiEEARQ+IQQBFD4hBAEUPxEEARQ/EQQBLAdPiEEARQ/EQQBrMdPBEEA0/ETQRDAdPxEEAQwHT8RBAEUQI9+8HxPbiaA4ieCAiiAAih+IiiAAiiA4ieCAiiAAiiAAiiAAiiA4ieCAiiAAiiAAiiAAiiA4ieCAiiAAiiAAiiAAiiA4ieCAiiAAiiAAhgPoJiJoAAKoAAigAIogAKIAAqgAAYCKGIiKIACKIAIoAAKoAAigAIogIEAipcICqAACiACKIACKIAIoAAKoAAigAIogAKIAAqgAAogAiiAAiiACKAACqAAIoAIoAAigAigAAqgACKAAiiAAogACqAACiACKIACKIAIoAAKoAAigAIogAKIAAqgAAogAiiAAiiACKAACqAAIoACKIACiACKoPgJoAAKoAAKoAAKoAAKoAAKoAAKoAiKnwAKoAAKoAAKoAAKoAAKoAAKoACKoPgJoAAKoAAKoAAKoACKoPgJoAAKoAAKoAAKoACKoPgJoAAKoAAKoAAKYD6AIih+AiiAAogACqAAFgMoguIngAKYDqAIip8ACmA6gCIofgIogAIogAigABYDWI+g2AmgAMYDWI2g0AmgAApgMoIiJ4ACKIDJCAqcAAqgACYjKG4CKIACmAuhqAmgAApgMoKCJoACKIDJCLqfAAqgAOZC6F4CKIACmAuh+wigAApgLoTuIYACKIC5ENpfAAVQAFMxtLMACqAApmJoTwEUQAEEARRAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEBAAI0ACCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAKc9wUAIELiSQyF+gAAAABJRU5ErkJggg==",
  "base64"
);

function isWeaveImage(value: unknown): value is weave.WeaveImage {
  return (
    typeof value === "object" &&
    value !== null &&
    "_weaveType" in value &&
    (value as { _weaveType?: unknown })._weaveType === "Image"
  );
}

async function imageResultToWeavePng(item: unknown): Promise<weave.WeaveImage> {
  if (isWeaveImage(item)) {
    return item;
  }

  const image = item as { b64_json?: string; url?: string };
  if (image.b64_json) {
    return weave.weaveImage({ data: Buffer.from(image.b64_json, "base64") });
  }

  if (image.url) {
    const response = await fetch(image.url);
    if (!response.ok) {
      throw new Error(`Failed to download generated image: ${response.status}`);
    }
    return weave.weaveImage({ data: Buffer.from(await response.arrayBuffer()) });
  }

  throw new Error("OpenAI image response did not include b64_json, url, or WeaveImage data.");
}

// =============================================================================
// SECTION 1: PNG Image Tracing
// =============================================================================
printSection("1. 画像 (PNG) - TypeScript 対応: weaveImage");

const createSamplePng = weave.op(function createSamplePng() {
  return weave.weaveImage({ data: SAMPLE_VISIBLE_PNG });
});

const sampleImage = await createSamplePng();
console.log(`Sample PNG recorded as Weave media: ${sampleImage._weaveType}, bytes=${sampleImage.data.length}`);

const generateImage = weave.op(async function generateImage(prompt: string) {
  const response = await getOpenAIClient().images.generate({
    model: getImageModelName(),
    prompt,
    size: "1024x1024",
  });
  const firstImage = response.data?.[0];
  if (!firstImage) {
    throw new Error("OpenAI image response did not include any image data.");
  }
  return imageResultToWeavePng(firstImage);
});

if (process.env.RUN_EXPENSIVE_MULTIMODAL === "1") {
  const generated = await generateImage("a small robot watering a flower in a lab, pixel art style");
  console.log(`Generated PNG recorded as Weave media: ${generated._weaveType}`);
} else {
  console.log("RUN_EXPENSIVE_MULTIMODAL=1 を設定すると OpenAI 画像生成の PNG trace 例を実行します。");
}

// =============================================================================
// SECTION 2: JPG Image Tracing
// =============================================================================
// TypeScript 対応: 未対応
// Python 版:
//   def create_sample_jpg_image() -> Annotated[bytes, Content[Literal["jpg"]]]:
//       ...
//
// TypeScript SDK の ImageType は現時点で "png" のみです。
// JPG bytes を Python 版と同じ media preview として trace する helper はありません。

// =============================================================================
// SECTION 3: MP3 Audio Tracing
// =============================================================================
// TypeScript 対応: 未対応
// Python 版:
//   def text_to_speech(...) -> Annotated[bytes, Content[Literal["mp3"]]]:
//       ...
//
// OpenAI TTS は TypeScript でも response_format: "mp3" を指定できます。
// ただし TypeScript SDK の AudioType は現時点で "wav" のみなので、
// MP3 bytes を Python 版と同じ media preview として trace する helper はありません。

// =============================================================================
// SECTION 4: WAV Audio Tracing
// =============================================================================
printSection("4. 音声 (WAV) - TypeScript 対応: weaveAudio");

const textToSpeechWav = weave.op(async function textToSpeechWav(text: string) {
  const response = await getOpenAIClient().audio.speech.create({
    model: getTtsModelName(),
    voice: "alloy",
    input: text,
    response_format: "wav",
  });
  return weave.weaveAudio({ data: Buffer.from(await response.arrayBuffer()), audioType: "wav" });
});

if (process.env.RUN_EXPENSIVE_MULTIMODAL === "1") {
  const audio = await textToSpeechWav(
    "Weights and Biases Weave helps you trace, evaluate, and monitor LLM applications."
  );
  console.log(`Generated WAV recorded as Weave media: ${audio._weaveType}`);
} else {
  console.log("RUN_EXPENSIVE_MULTIMODAL=1 を設定すると OpenAI TTS の WAV trace 例を実行します。");
}

// =============================================================================
// SECTION 5: MP4 Video Tracing
// =============================================================================
// TypeScript 対応: 未対応
// Python 版:
//   def generate_video(...) -> Annotated[bytes, Content[Literal["mp4"]]]:
//       ...
//
// TypeScript SDK には MP4 用の media helper がありません。
// この教材では TypeScript 版の MP4 trace 実行例は用意していません。

// =============================================================================
// SECTION 6: PDF Tracing
// =============================================================================
// TypeScript 対応: 未対応
// Python 版:
//   def fetch_arxiv_pdf(...) -> Annotated[bytes, Content[Literal["pdf"]]]:
//       ...
//
// TypeScript SDK には PDF 用の Content annotation / media helper がありません。
// PDF を扱う場合は URL や byte size などを通常 object として trace する代替になります。

// =============================================================================
// SECTION 7: HTML Tracing
// =============================================================================
// TypeScript 対応: 未対応
// Python 版:
//   def fetch_html(...) -> Annotated[bytes, Content[Literal["html"]]]:
//       ...
//
// TypeScript SDK には HTML 用の Content annotation / media helper がありません。
// HTML を扱う場合は preview text や metadata object として trace する代替になります。

// =============================================================================
// COMPLETE
// =============================================================================
printSection("Multimodal Data Tracing Demo Complete!");
console.log(`
まとめ:
- Python: png / jpg / mp3 / mp4 / pdf / html を Content annotation で media preview として trace
- TypeScript: PNG は weaveImage で media trace
- TypeScript: WAV は weaveAudio で media trace
- TypeScript: JPG / MP3 / MP4 / PDF / HTML は、この教材ではコメントで未対応理由のみ説明
`);
